src/topology/algebra/ordered.lean

```diff
@@ -1034,35 +1100,87 @@ lemma cinfi_of_cinfi_of_monotone_of_continuous {f : α → β} {g : γ → α}
 by rw [infi, cInf_of_cInf_of_monotone_of_continuous Mf Cf
   (λ h, range_eq_empty.1 h ‹_›) H, ← range_comp]; refl
 
+/-- A "continuous induction principle" for a closed interval: if a set `s` meets `[a, b]`
+on a closed subset, contains the left endpoint of a closed interval `[a, b]`, and the set
+`s ∩ [a, b)` has no maximal point, then `b ∈ s`. -/
+lemma is_closed.mem_of_ge_of_forall_exists_gt {a b : α} {s : set α} (hs : is_closed (s ∩ Icc a b))
+  (ha : a ∈ s) (hab : a ≤ b) (hgt : ∀ x ∈ s ∩ Ico a b, (s ∩ Ioc x b).nonempty) :
+  b ∈ s :=
+begin
+  let S := s ∩ Icc a b,
+  replace ha : a ∈ S, from ⟨ha, left_mem_Icc.2 hab⟩,
+  have Sne: S ≠ ∅, from ne_empty_of_mem ha,
+  have Sbd : bdd_above S, from ⟨b, λ z hz, hz.2.2⟩,
+  let c := Sup (s ∩ Icc a b),
+  have c_mem : c ∈ S, from cSup_mem_of_is_closed Sne hs Sbd,
+  have c_le : c ≤ b, from cSup_le Sne (λ x hx, hx.2.2),
+  cases eq_or_lt_of_le c_le with hc hc, from hc ▸ c_mem.1,
+  exfalso,
+  rcases hgt c ⟨c_mem.1, c_mem.2.1, hc⟩ with ⟨x, xs, cx, xb⟩,
+  exact not_lt_of_le (le_cSup Sbd ⟨xs, le_trans (le_cSup Sbd ha) (le_of_lt cx), xb⟩) cx
+end
+
+/-- A "continuous induction principle" for a closed interval: if a set `s` meets `[a, b]`
+on a closed subset, contains the left endpoint of a closed interval `[a, b]`, and for any
+`a ≤ x < y ≤ b`, `x ∈ s`, the set `s ∩ (x, y]` is not empty, then `[a, b] ⊆ s`. -/
+lemma is_closed.Icc_subset_of_forall_exists_gt {a b : α} {s : set α} (hs : is_closed (s ∩ Icc a b))
+  (ha : a ∈ s) (hab : a ≤ b) (hgt : ∀ y ∈ Icc a b, ∀ x ∈ s ∩ Ico a y,  (s ∩ Ioc x y).nonempty) :
+  Icc a b ⊆ s :=
+begin
+  assume y hy,
+  have : is_closed (s ∩ Icc a y),
+  { suffices : s ∩ Icc a y = s ∩ Icc a b ∩ Icc a y,
+    { rw this, exact is_closed_inter hs is_closed_Icc },
+    rw [inter_assoc],
+    congr,
+    exact (inter_eq_self_of_subset_right $ Icc_subset_Icc_right hy.2).symm },
+  exact is_closed.mem_of_ge_of_forall_exists_gt this ha hy.1 (hgt y hy)
+end
+
 section densely_ordered
 
 variables [densely_ordered α] {a b : α}
 
+/-- A "continuous induction principle" for a closed interval: if a set `s` meets `[a, b]`
+on a closed subset, contains the left endpoint of a closed interval `[a, b]`, and for any
```

Comment:
`the left endpoint of a closed interval `[a, b]`` is strange, as you have already mentioned `[a,b]` on the previous line.

Reply:
Fixed.